### PR TITLE
[IMP] use file_path and file_open

### PR DESCRIPTION
**Pull request by mart-e**
Deprecate the usage of `get_resource_path`/`get_module_resource`. Using `file_path` has the same features with additional security checks.

Replace all the deprecated calls to `open` or `get_resource_path` with the `file_open` and `file_path` method.

Differences between `get_resource_path` and `file_path`:
- `get_resource_path` returns False if the file does not exsits, `file_path` raises a `FileNotFoundError`
- `file_path` ensures the file is within the addons-path

Difference between `open` and `file_open`
- `file_open` ensures the file is within the addons-path

### CODE REVIEW

**Review by Julien00859 — APPROVED** [review]
ir_http

**Comment by tde-banana-odoo** [comment]
> ir_http

Is it a game ?

-> ir_model

**Review by odony — APPROVED** [review]
nice cleanup!

@robodoo override=ci/security
B: https://runbot.odoo.com/runbot/build/51801285
F: 
- 24 instances of `os` references, most `os.path.join` in controlled situations before calling `file_path` or `file_open` ; one case of `os.sep`
- 3 `subprocess` callsites touched, in controlled situations with fixed paths and `file_path`
- 3 `stream.from_path` with controlled values

File: odoo/tools/misc.py [hunk 1]
```diff
@@ -178,17 +180,24 @@ def file_path(file_path, filter_ext=('',), env=None):
...
                  os.path.normpath(os.path.normcase(os.path.join(parent_path, normalized_path))))
         if fpath.startswith(parent_path) and os.path.exists(fpath):
+            if create:
+                return os.path.join(fpath, file_name)
             return fpath
 
     raise FileNotFoundError("File not found: " + file_path)
 
-def file_open(name, mode="r", filter_ext=None, env=None):
+def file_open(name, mode="r", filter_ext=None, env=None, create=False):
```
[the remark this answers is not in the document]
Comment (Xavier-Do) [reply odoo/tools/misc.py:200]: It makes sens in odoo but on runbot, the create option would be useful. We need to create file in the runbot/static/... directory.

**Comment by rco-odoo** [comment]
@mart-e I didn't mean to be rude with you.  I should have rephrased my comment a little bit.
By "what a terrible API", what I meant is that it really doesn't feel right to me.
The parameter `create` really isn't necessary in `file_path()`; in `file_open()` you may do this instead:
```py
    if create:
        dirname, filename = os.path.split(name)
        path = os.path.join(file_path(dirname, env=env), filename)
    else:
        path = file_path(name, filter_ext=filter_ext, env=env)
```
Well, now that I look at it, we've lost the `filter_ext` feature.  We might need to factor out that feature in a separate function... :thinking:

**Comment by rco-odoo** [comment]
By the way, it looks like `file_open(..., create=True)` enables to create a file inside an addon directory.  Isn't that risky?

**Comment by mart-e** [comment]
> By the way, it looks like `file_open(..., create=True)` enables to create a file inside an addon directory. Isn't that risky?

Yes, I don't see any valid reason to allow that. It's actually only useful in the `file_open_temporary_directory` scenario.

File: addons/web/controllers/binary.py [hunk 2]
```diff
@@ -255,9 +253,9 @@ def company_logo(self, dbname=None, **kw):
                             response_class=Response,
                         )
                     else:
-                        response = http.Stream.from_path(placeholder('nologo.png')).get_response()
+                        response = http.Stream.from_path(file_path('web/static/img/nologo.png')).get_response()
             except Exception:
-                response = http.Stream.from_path(placeholder(imgname + imgext)).get_response()
+                response = http.Stream.from_path(file_path(f'web/static/img/{imgname}{imgext}')).get_response()
```
[the remark this answers is not in the document]
Comment (mart-e) [reply addons/web/controllers/binary.py:258]: no here the extension can be changed to svg above

File: odoo/tools/translate.py [hunk 3]
```diff
@@ -1239,9 +1239,13 @@ def load_file(self, filepath, lang, xmlids=None):
...
-        with file_open(filepath, mode='rb') as fileobj:
-            fileformat = os.path.splitext(filepath)[-1][1:].lower()
-            self.load(fileobj, fileformat, lang, xmlids=xmlids)
+        try:
+            with file_open(filepath, mode='rb') as fileobj:
+                _logger.info('loading base translation file %s for language %s', filepath, lang)
+                fileformat = os.path.splitext(filepath)[-1][1:].lower()
+                self.load(fileobj, fileformat, lang, xmlids=xmlids)
+        except FileNotFoundError:
+            pass
```
Comment (KangOl) [inline odoo/tools/translate.py:1248]:
```suggestion
        with suppress(FileNotFoundError), file_open(filepath, mode='rb') as fileobj:
            _logger.info('loading base translation file %s for language %s', filepath, lang)
            fileformat = os.path.splitext(filepath)[-1][1:].lower()
            self.load(fileobj, fileformat, lang, xmlids=xmlids)
```

**Review by KangOl — CHANGES_REQUESTED** [review]

**Comment by mart-e** [comment]
I have moved the `file_open_temporary_directory` into odoo/odoo#137695 as it's the part where discussion remains but it's only for details. This way we can already merge the file_path part.

@robodoo r+ rebase-ff

**Comment by mart-e** [comment]
@robodoo r+

**Comment by mart-e** [comment]
@robodoo r+ (hoping no longer having conflicts)

**Comment by Armin-FalDiS** [comment]
Hello @mart-e 
This seems to have broken support for all non-png icons. Is there a reason for this or was this side-effect unintentional?

**Comment by mart-e** [comment]
@Armin-FalDiS can you be a bit more specific? This commit changed a lot of things and was merged 2 years ago.

edit: oh I see, the ir_ui_menu.py part. Hmm it’s unintentional. We only have .png icons but we could support other image format if needed. Don’t hesitate to make a PR to patch this.

**Pull request closed without merging**